drivers/serial/usart_gd32.c

```diff
@@ -5,35 +5,90 @@
 
 #define DT_DRV_COMPAT gd_gd32_usart
 
+#include <errno.h>
 #include <drivers/uart.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(usart_gd32);
 
 struct gd32_usart_config {
 	uint32_t reg;
 	uint32_t rcu_periph_clock;
+	uint32_t parity;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_config_func_t irq_config_func;
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
 };
 
 struct gd32_usart_data {
 	uint32_t baud_rate;
+	uart_irq_callback_user_data_t user_cb;
+	void *user_data;
 };
 
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static void usart_gd32_isr(const struct device *dev)
+{
+	struct gd32_usart_data *const data = dev->data;
+
+	if (data->user_cb) {
+		data->user_cb(dev, data->user_data);
+	}
+}
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
```

Comment:
This should be moved inside same block related to interrupt feature

Reply:
I'm prefer to split the API implement and utility function.

Reply:
UART_ASYNC_API and Power Management will use usart_gd32_isr.